File: java/java-advent-of-code-project-002/solve_advent_of_code_with_java_002/src/main/java/com/peter_burbery/solve_advent_of_code_with_java_002/year_2024/day_004/XMASSolver.java
package com.peter_burbery.solve_advent_of_code_with_java_002.year_2024.day_004;

import java.io.*;
import java.util.*;

public class XMASSolver {

    public static void main(String[] args) throws IOException {
        // Read the file and store the grid
        String filePath = "\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-004\\input.txt";
        List<String> gridList = new ArrayList<>();
        
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            String line;
            while ((line = br.readLine()) != null) {
                gridList.add(line);
            }
        }
        
        String[] grid = gridList.toArray(new String[0]);
        int rows = grid.length;
        int cols = grid[0].length();

        // Count occurrences of "XMAS"
        int xmasCount = countXMAS(grid, rows, cols);
        System.out.println("Count of XMAS: " + xmasCount);

        // Count all X-MAS patterns
        int xmasPatterns = countAllXMASPatterns(grid, rows, cols);
        System.out.println("Total X-MAS patterns: " + xmasPatterns);
    }

    // Count occurrences of "XMAS" in all directions
    public static int countXMAS(String[] grid, int rows, int cols) {
        String targetWord = "XMAS";
        int count = 0;
        int[][] directions = {
            {0, 1},  // Right
            {1, 0},  // Down
            {1, 1},  // Diagonal-right-down
            {1, -1}, // Diagonal-left-down
            {0, -1}, // Left
            {-1, 0}, // Up
            {-1, -1},// Diagonal-left-up
            {-1, 1}  // Diagonal-right-up
        };

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                for (int[] dir : directions) {
                    if (checkWord(grid, rows, cols, r, c, dir[0], dir[1], targetWord)) {
                        count++;
                    }
                }
            }
        }
        return count;
    }

    // Check if a word exists in a given direction
    public static boolean checkWord(String[] grid, int rows, int cols, int x, int y, int dx, int dy, String targetWord) {
        for (int i = 0; i < targetWord.length(); i++) {
            int nx = x + i * dx;
            int ny = y + i * dy;
            if (nx < 0 || ny < 0 || nx >= rows || ny >= cols || grid[nx].charAt(ny) != targetWord.charAt(i)) {
                return false;
            }
        }
        return true;
    }

    // Count all X-MAS patterns
    public static int countAllXMASPatterns(String[] grid, int rows, int cols) {
        int count = 0;
        for (int r = 1; r < rows - 1; r++) {
            for (int c = 1; c < cols - 1; c++) {
                char center = grid[r].charAt(c);
                char topLeft = grid[r - 1].charAt(c - 1);
                char topRight = grid[r - 1].charAt(c + 1);
                char bottomLeft = grid[r + 1].charAt(c - 1);
                char bottomRight = grid[r + 1].charAt(c + 1);

                if (center == 'A') {
                    // Pattern 1: M.S
                    if (topLeft == 'M' && topRight == 'S' && bottomLeft == 'M' && bottomRight == 'S') {
                        count++;
                    }
                    // Pattern 2: S.M
                    else if (topLeft == 'S' && topRight == 'M' && bottomLeft == 'S' && bottomRight == 'M') {
                        count++;
                    }
                    // Pattern 3: M.M
                    else if (topLeft == 'M' && topRight == 'M' && bottomLeft == 'S' && bottomRight == 'S') {
                        count++;
                    }
                    // Pattern 4: S.S
                    else if (topLeft == 'S' && topRight == 'S' && bottomLeft == 'M' && bottomRight == 'M') {
                        count++;
                    }
                }
            }
        }
        return count;
    }
}

File: other-programming-languages/Scala-002/2024/2024-004/XMASSolver.scala
import scala.io.Source

object XMASSolver {

  def main(args: Array[String]): Unit = {
    val filePath = "\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-004\\input.txt"
    
    // Read the file and store the grid
    val grid = Source.fromFile(filePath).getLines().toArray
    val rows = grid.length
    val cols = grid(0).length

    // Count occurrences of "XMAS"
    val xmasCount = countXMAS(grid, rows, cols)
    println(s"Count of XMAS: $xmasCount")

    // Count all X-MAS patterns
    val xmasPatterns = countAllXMASPatterns(grid, rows, cols)
    println(s"Total X-MAS patterns: $xmasPatterns")
  }

  // Count occurrences of "XMAS" in all directions
  def countXMAS(grid: Array[String], rows: Int, cols: Int): Int = {
    val targetWord = "XMAS"
    val directions = Seq(
      (0, 1),   // Right
      (1, 0),   // Down
      (1, 1),   // Diagonal-right-down
      (1, -1),  // Diagonal-left-down
      (0, -1),  // Left
      (-1, 0),  // Up
      (-1, -1), // Diagonal-left-up
      (-1, 1)   // Diagonal-right-up
    )

    // Helper function to check if the word exists in a given direction
    def checkWord(x: Int, y: Int, dx: Int, dy: Int): Boolean = {
      (0 until targetWord.length).forall { i =>
        val nx = x + i * dx
        val ny = y + i * dy
        nx >= 0 && ny >= 0 && nx < rows && ny < cols && grid(nx)(ny) == targetWord(i)
      }
    }

    // Traverse the grid to count occurrences
    var count = 0
    for {
      r <- 0 until rows
      c <- 0 until cols
      (dx, dy) <- directions
    } if (checkWord(r, c, dx, dy)) count += 1

    count
  }

  // Count all X-MAS patterns
  def countAllXMASPatterns(grid: Array[String], rows: Int, cols: Int): Int = {
    // Traverse the grid ensuring bounds for a 3x3 X-MAS pattern
    def checkPattern(center: Char, topLeft: Char, topRight: Char, bottomLeft: Char, bottomRight: Char): Boolean = {
      (center == 'A' && (
        (topLeft == 'M' && topRight == 'S' && bottomLeft == 'M' && bottomRight == 'S') || // Pattern 1: M.S
        (topLeft == 'S' && topRight == 'M' && bottomLeft == 'S' && bottomRight == 'M') || // Pattern 2: S.M
        (topLeft == 'M' && topRight == 'M' && bottomLeft == 'S' && bottomRight == 'S') || // Pattern 3: M.M
        (topLeft == 'S' && topRight == 'S' && bottomLeft == 'M' && bottomRight == 'M')    // Pattern 4: S.S
      ))
    }

    var count = 0
    for {
      r <- 1 until rows - 1
      c <- 1 until cols - 1
    } {
      val center = grid(r)(c)
      val topLeft = grid(r - 1)(c - 1)
      val topRight = grid(r - 1)(c + 1)
      val bottomLeft = grid(r + 1)(c - 1)
      val bottomRight = grid(r + 1)(c + 1)

      if (checkPattern(center, topLeft, topRight, bottomLeft, bottomRight)) count += 1
    }

    count
  }
}
